File: Sintaxis I/Tema - Asignacion.cs
// ═════════════════════════════════════
//          Operadores Simbolos
// ═════════════════════════════════════


// ┌───────────────────────────────────┐
// │                ..                 │
// └───────────────────────────────────┘
int[] numbers = new[] { 0, 10, 20, 30, 40, 50 };
int amountToDrop = numbers.Length / 2;

int[] rightHalf = numbers[amountToDrop..];
Display(rightHalf); // output: 30 40 50

int[] leftHalf = numbers[..amountToDrop];
Display(leftHalf); // output: 0 10 20

int[] all = numbers[..];
Display(all); // output: 0 10 20 30 40 50

void Display<T>(IEnumerable<T> xs) => Console.WriteLine(string.Join(" ", xs));

// ┌───────────────────────────────────┐
// │                ?                  │
// └───────────────────────────────────┘
// Sin Valor Predeterminado Sera Null
// Declaramos un int nullable
int? numero = null;

Console.WriteLine(numero.HasValue); // Imprime: False (no tiene valor)
Console.WriteLine(numero); // Imprime: (nada, porque es null)

numero = 42;
Console.WriteLine(numero.HasValue); // Imprime: True (tiene valor)
Console.WriteLine(numero); // Imprime: 42

// ┌───────────────────────────────────┐
// │               ??                  │
// └───────────────────────────────────┘
// Si es Null , Selecciona Valor por Defecto

string nombre = null;
string resultado = nombre ?? "Respuesta Predeterminada";

Console.WriteLine($"Bienvenido, {resultado}"); // Imprime: Bienvenido, Invitado

// ═════════════════════════════════════
//        Operadores Especiales
// ═════════════════════════════════════


// ┌───────────────────────────────────┐
// │            Ternario               │
// └───────────────────────────────────┘
int edad = 20;
string mensaje = (edad >= 18) ? "Mayor de edad" : "Menor de edad";
Console.WriteLine(mensaje); // Imprime: Mayor de edad

bool estaLloviendo = true;
Console.WriteLine(estaLloviendo ? "Lleva un paraguas" : "Disfruta del sol");

// ┌───────────────────────────────────┐
// │          Deconstruccion           │
// └───────────────────────────────────┘

var persona1 = (Nombre: "Ana", Edad: 25);
var persona2 = (Nombre: "Luis", Edad: 30, Altura: 1.78);

// [Tuplas] [Usar var simplifica]
(string nombre1, int edad1) = persona1;
(string nombre2, _, double altura2) = persona2;
(string nombre3, int edad3) = ("María", 28);

Console.WriteLine($"Nombre: {nombre1} Edad: {edad1}");
Console.WriteLine($"Nombre: {nombre2}, Altura: {altura2}");
Console.WriteLine($"Nombre: {nombre3}, Edad: {edad3}");

public class Punto
{
    public int X { get; }
    public int Y { get; }

    public Punto(int x, int y)
    {
        X = x;
        Y = y;
    }

    public void Deconstruct(out int x, out int y)
    {
        x = X;
        y = Y;
    }
}

/*var punto = new Punto(3, 4);*/
/*(var x, var y) = punto;*/
/*Console.WriteLine($"X: {x}, Y: {y}");*/

File: Sintaxis I/Tema - Asignacion.js
// ┌───────────────────────────────────┐
// │                &&                 │
// └───────────────────────────────────┘

let nombre1 = "Ana";
let edad1 = 25;

// Usar && para verificar si ambas condiciones son truthy
let puedeEntrar = nombre1 && edad1 >= 18;
console.log(puedeEntrar); // Imprime: true (porque "Ana" es truthy y 25 >= 18 es true)

// Ejemplo con un valor falsy
let usuario = "";
let contraseña = "12345";
let loginValido = usuario && contraseña;
console.log(loginValido); // Imprime: "" (devuelve el primer valor falsy: usuario)

// Ejemplo con cortocircuito
let mensaje = nombre && "Bienvenida, " + nombre;
console.log(mensaje); // Imprime: "Bienvenida, Ana"

// ┌───────────────────────────────────┐
// │                ||                 │
// └───────────────────────────────────┘
let nombreUsuario = "";
let nombrePorDefecto = "Invitado";

// Usar || para asignar un valor por defecto
let nombreFinal = nombreUsuario || nombrePorDefecto;
console.log(nombreFinal); // Imprime: "Invitado" (porque "" es falsy)

// Ejemplo con valores truthy
let colorFavorito = "azul";
let colorAlternativo = "verde";
let colorElegido = colorFavorito || colorAlternativo;
console.log(colorElegido); // Imprime: "azul" (devuelve el primer valor truthy)

// Ejemplo con cortocircuito
let configuracion = null;
let configPorDefecto = { tema: "oscuro" };
let configFinal = configuracion || configPorDefecto;
console.log(configFinal); // Imprime: { tema: "oscuro" }
// ┌───────────────────────────────────┐
// │                ??                 │
// └───────────────────────────────────┘

// ┌───────────────────────────────────┐
// │                 ?                 │
// └───────────────────────────────────┘

let usuario1 = {
  nombre: "Ana",
  direccion: {
    ciudad: "Madrid",
  },
};

console.log(usuario1?.direccion?.ciudad); // Imprime: "Madrid"
console.log(usuario1?.contacto?.telefono); // Imprime: undefined (no genera error)

// ═════════════════════════════════════
//        Operadores Especiales
// ═════════════════════════════════════

// ┌───────────────────────────────────┐
// │            Ternario               │
// └───────────────────────────────────┘
const resultado = a > b ? "mayor" : "menor"; // "mayor" (if corto)

function saludar(persona, sexo) {
  return sexo == "m"
    ? `Bienvenido a EDteam, ${persona}`
    : `Bienvenida a EDteam, ${persona}`;
}

console.log(saludar("Beto", "m"));
console.log(saludar("Alexys", "m"));
console.log(saludar("Juan Carlos", "m"));
console.log(saludar("Maria", "f"));
